log/log.go

```diff
@@ -1,159 +1,45 @@
-)
-
-// Init initializes the logger. This is required to use logging correctly
-// component is the name of the component being used to log messages.
-// Typically this is your application name.
-//
-// keyValuePairs are key/value pairs to be used with all logs in the future
-func Init(component string, keyValuePairs ...interface{}) {
-	InitWithOptions(component, nil, keyValuePairs...)
-}
-
-// MustInit calls Init and panics if it returns an error
-func MustInit(component string, keyValuePairs ...interface{}) {
-	Init(component, keyValuePairs...)
+// NewLogger creates a logger with the provided key value pairs
+func NewLogger(component string, keyValuePairs ...interface{}) logr.Logger {
+	return NewLoggerWithOptions(component, nil, keyValuePairs...)
 }
 
-// InitWithOptions inits the logger with the provided opts
-func InitWithOptions(component string, opts []Option, keyValuePairs ...interface{}) {
-	mtx.Lock()
-	defer mtx.Unlock()
+// NewLoggerWithOptions creates a logger with the provided opts and key value pairs
+func NewLoggerWithOptions(component string, opts *Options, keyValuePairs ...interface{}) logr.Logger {
+	sinkOpts := Options{Writer: os.Stdout, LogLevel: 0}
 
-	ll := NewLogger(component, defaultOutput, 0, JSONEncoder{}, keyValuePairs...)
-
-	for _, opt := range opts {
-		opt(ll)
+	if opts != nil {
+		sinkOpts.Writer = opts.Writer
+		sinkOpts.LogLevel = opts.LogLevel
 	}
 
-	// don't lock because we already have a lock
-	useLogger(ll)
-}
-
-// MustInitWithOptions calls InitWithOptions and panics if an error is returned
-func MustInitWithOptions(component string, opts []Option, keyValuePairs ...interface{}) {
-	InitWithOptions(component, opts, keyValuePairs...)
-}
-
-// GetLogger returns the root logger used for logging
-func GetLogger() logr.Logger {
-	return logger
-}
-
-// UseLogger bypasses the requirement for Init and sets the logger to l
-func UseLogger(l logr.Logger) {
-	mtx.Lock()
-	defer mtx.Unlock()
-	useLogger(l)
-}
-
-// useLogger sets the logger to l without mtx.Lock()
-// To use mtx.Lock see UseLogger
-func useLogger(l logr.Logger) {
-	logger = l
+	return logr.New(
+		sink.NewLogSink(
+			component,
+			sinkOpts.Writer,
+			sink.Verbosity(sinkOpts.LogLevel),
+			sink.JSONEncoder{},
+			keyValuePairs...,
+		),
+	)
 }
 
-// Info logs a non-error message with the given key/value pairs as context.
-//
-// The msg argument should be used to add some constant description to
-// the log line.  The key/value pairs can then be used to add additional
-// variable information.  The key/value pairs should alternate string
-// keys and arbitrary values.
-func Info(msg string, keysAndValues ...interface{}) {
-	mtx.RLock()
-	defer mtx.RUnlock()
-	logger.Info(msg, keysAndValues...)
-}
+// NewLoggerV calls the logr.Logger V() method while updating the internal sink (if possible)
+// to be compatible with the new level
+func NewLoggerV(logger logr.Logger, level int) logr.Logger {
```

Comment:
I think `NewLoggerV` is misleading because this modifies the underlying `Sink` for all loggers that use the same Sink.

If we want to support setting the log level after creating the sink, then something like `log.SetV(logr.Logger, int) bool` might be more fitting. I don't know if this is actually a use-case that we have, though.

Reply:
My understanding is that the sink's verbosity is set once during application startup from some kind of configuration option (CLI arguments, environment variables, configuration files, ...) and then never changed. The `V()` method of the logger sets the verbosity of the emitted lines and is used throughout the code to emit log entries with different verbosity that are sent to the output depending on the verbosity of the sink.

Forcing the sink's verbosity to another value for a part of the code seems like an anti-pattern to me and would either modify the globally-used sink or would need to introduce a scoped sink.

Having different verbosities for different parts of an application (which Log4j and similar libraries have) seems not to be something supported by the current approach.